chiselFrontend/src/main/scala/chisel3/core/MultiClock.scala

```diff
@@ -0,0 +1,25 @@
+// See LICENSE for license details.
+
+package chisel3.core
+
+import scala.language.experimental.macros
+
+import chisel3.internal._
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl._
+import chisel3.internal.sourceinfo.{SourceInfo}
+
+private[chisel3] final case class ClockAndReset(clock: Clock, reset: Bool)
+
+object withClockAndReset {  // scalastyle:ignore object.name
+  def apply[T](clock: Clock, reset: Bool)(block: => T)(implicit sourceInfo: SourceInfo): T = {
+    // Save parentScope
+    val parentScope = Builder.currentClockAndReset
+    Builder.currentClockAndReset = Some(ClockAndReset(clock, reset))
+    val res = block // execute block
+    // Return to old scope
+    Builder.currentClockAndReset = parentScope
+    res
```

Comment:
I can see why returning a value might be useful, but there's also the pathological case of trying to return a function defined in the withClockAndReset scope. Invoking it outside would cause it to execute in the outer clock and reset scope, which might not be what is expected.
That being said, I think that use case would look obviously wrong, and even without this feature a user could just overwrite a var in an outer scope.
Are there other edge cases that we haven't thought of yet that might lead to unintuitive behavior?